### DISCUSSION TITLE
Zephyr-based Border Router

### DISCUSSION THREAD

**Discussion by beriberikix** [openthread/openthread#7482] [Ideas]
👋 Hi all! I've been looking into what it would take to implement a Border Router based on [Zephyr](https://www.zephyrproject.org/) and would love some feedback.

### Background

Our company [golioth.io](http://golioth.io/) provides device management capabilities for microcontroller-class devices that are supported by Zephyr. That allows us to support, for example, secure end-to-end software updates with secure boot for 100s of host MCUs and various forms of connectivity including Wi-Fi, Ethernet & Cellular. We also provide messaging, storage & data forwarding services across those ranges of hardware and connectivity. We're seeing an increase of interest to add Thread support to Golioth for both commercial and non-commercial use cases. The initial implementations will primarily be for managing end devices with Golioth but we'd like to support border routers as well.

Additionally, we're active contributors and sponsors of Zephyr so we're interested in seeing the project grow to expand new use cases 😀

### What exists today
 
OpenThread is the [official](https://docs.zephyrproject.org/latest/reference/networking/thread.html) Thread implementation for Zephyr & there are already examples for [NCP & RCP](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/net/openthread/coprocessor#building-and-running). Zephyr has a suite of native networking implementations, and some app layer protocols such as [DNS/mDNS](https://docs.zephyrproject.org/latest/reference/networking/dns_resolve.html). DNS-SD is under active [development](https://github.com/zephyrproject-rtos/zephyr/issues/38846).

Outside of Zephyr there is an MCU BR [implementation from Espressif](https://openthread.io/guides/border-router/espressif-esp32) based on their ESP-IDF/FreeRTOS. There is also of course [ot-br-posix](https://github.com/openthread/ot-br-posix), which builds upon Unix libraries like dnsmasq and tayga.

### What's missing

Zephyr currently lacks implementations for several of the app layer networking protocols, such as a [DHCP server](https://github.com/zephyrproject-rtos/zephyr/issues/38284) (of any kind,) NAT64 & DNS64, as well as others suggested in the [Thread Border Router Best Practices](https://www.threadgroup.org/Portals/0/documents/support/ThreadBorderRouterBestPractices_2530_1.pdf).)  There may be additional features or glue code missing (ex. integrating the Thread Border Agent.)

### Next steps

I plan on trying to build an example that uses the Espressif BR but instead of using the Espressif RCP, use a Zephyr-based RCP running Zephyr (probably an nRF52.)  I believe that should work relatively easily, but I'll only know once I try!

### Suggestions welcome!

Are there any suggestions on how to move forward on an implementation in Zephyr? What should be in a minimal BR, what libraries to use, any gotchas to keep in mind?

Many thanks in advance! 🙏

**Comment by jwhui** [comment]
Hey @beriberikix! Great to hear about your efforts to enable OpenThread Border Router. Looks like you're already done your homework.

A Thread Border Router can provide many functions, including:

- IPv6 connectivity
- IPv4 connectivity
- DNS-based service discovery
  - Service Registration Protocol
  - Advertising Proxy
  - Discovery Proxy
- DNS recursive resolver
- DNS authoritative server
- Thread-over-Infrastructure

Depending on your specific use case, you may or may not depend on all of the above mentioned features.

We are actively working on enabling all of the mentioned features in the OpenThread project. The more Thread-specific functions are implemented within the OpenThread core library, for example:

- IPv6 connectivity in https://github.com/openthread/openthread/tree/main/src/core/border_router
- Service Registration Protocol in https://github.com/openthread/openthread/tree/main/src/core/net
- Thread-over-Infrastructure in https://github.com/openthread/openthread/tree/main/src/core/radio

A reason for doing so is to make it easier to enable OpenThread Border Router on RTOS-based platforms. Currently, the [ESP32 platform](https://openthread.io/guides/border-router/espressif-esp32) is the only RTOS-based example I'm aware of that integrates OpenThread Border Router functionality.

Also, in case you missed it, the Thread Group is [actively working to standardize the Border Router function](https://www.threadgroup.org/news-events/blog/ID/303/Expanding-the-IoT-Tapestry-Thread-Groups-2021-Wrap-Up#.YjAfFRDMI8Y):

> Another key effort has been standardizing the Border Router function, to make it simple for Thread devices to be discovered and communicate with those outside the Thread network. Like Wi-Fi access points, Thread Border Routers provide functionality such as routing traffic and connecting to the cloud, but are unique as they can be built into existing devices from any company – minimizing the need for additional dedicated hardware. This enhancement will be included in a new version of the Thread specification that will be backward compatible with the existing spec, and is planned to launch in Q1 2022. More details to unfold in the new year, including specifics about our certification program.

If you are interested in learning more in advance of the launch, I encourage you to join the Thread Group :)

**Reply by beriberikix** [reply]
Now that [1.3.0](https://www.threadgroup.org/ThreadSpec) is official made public:

1. Are the relevant features you listed above merged into `main`? Or alternatively, is the a list of 1.3 features supported?
2. Are you aware of any additional public RTOS examples besides Espressif? They've also seem to expand it significantly with their recent [SDK](https://github.com/espressif/esp-thread-br)

**Reply by jwhui** [reply]
> Are the relevant features you listed above merged into main? Or alternatively, is the a list of 1.3 features supported?

Yes, all Thread 1.3.0 features are included in main branch. The code used by the border router reference devices in the Thread Test Harness is here: https://github.com/openthread/ot-reference-release

> Are you aware of any additional public RTOS examples besides Espressif?

I'm not currently aware of any other public examples.
